veps/sig-compute/migration-prio-queue.md

```diff
@@ -0,0 +1,127 @@
+- Enable controllers to set priorities based on migration type.
+- Restrict user-set priorities via admission webhooks.
+- Ensure backward compatibility.
+
+## Non Goals
+
+- Change how migrations are triggered or processed.
+- Introduce new migration types.
+- Make migrations preemptible.
+
+## User Stories
+
+- As a cluster administrator, I want KubeVirt system-critical operations to proceed without delays from user actions.
+- As a VM owner, I want my hot plug operations processed efficiently, respecting system priorities.
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+A `priority` field will be added to the VMIM API, allowing various controllers to
+set priorities for migrations. The migration controller will use this field
+to order migrations in the priority queue, ensuring system-initiated migrations
+are processed before user-initiated ones. Running migrations will be get a reserved
+high priority (e.g., 1000).
+Controllers will set priorities based on the migration type, and admission webhooks will ensure users cannot
+set high priorities, protecting system operations.
+
+### Priority Order
+
+The priority order for migrations is as follows, with higher numerical values indicating higher priority:
+
+- Running Migrations: 1000 (Protected)
+- System-Critical Migrations: 100 - These include urgent tasks like node drains, evacuations, and KubeVirt upgrades. They’re critical for system health and need immediate attention.
+- User-Triggered Operations: 50 -  These are user-initiated changes, like hot plug operations.  They’re time-sensitive but should not affect system stability.
+- System Maintenance Migrations: 20 - Less urgent maintenance tasks, such as those started by the de-scheduler.
+- Manual VMIM Creation without priority or custom admin setting: This applies to admin-created VMIMs where no priority is set. They get the lowest priority by default.
+
+## API Examples
+
+```go
+type MigrationPriority string
+
+const (
+  SystemCritical    MigrationPriority = "system-critical"
+  UserTriggered     MigrationPriority = "user-triggered"
+  SystemMaintenance MigrationPriority = "system-maintenance"
+)
+
+const (
+  PriorityRunning           int32 = 1000
+  PrioritySystemCritical    int32 = 100
+  PriorityUserTriggered     int32 = 50
+  PrioritySystemMaintenance int32 = 20
+  PriorityDefault           int32 = 0
+)
+
+type VirtualMachineInstanceMigrationSpec struct {
+    // Priority of the migration, higher values indicate higher priority.
+    // +optional
+    Priority *MigrationPriority `json:"priority,omitempty"`
+[..]
+}
+```
+
+
+## Update/Rollback Compatibility
+
+ - VMIMs without `priority` default to 0 and the prioirity field is optional.
+
+## Functional Testing Approach
+
+- Test node drain, upgrades, hot plug, and mixed-priority scenarios. Make sure that the webhook is enforced.
+
+
+## Feature Lifecycle Phases
+
+### Alpha
+
+- Add `priority` to VMIM CRD behind a feature gate.
```

Comment:
What is the feature gate name?

Reply:
I think that this is truly an implementation detail.